### PR TITLE
[FINAL] Subscription offers support

### PR DESCRIPTION
**Pull request by vegaro**
- Adds `getPaymentDiscount` to get a signed discount. It will return `undefined` for incompatible iOS versions and Android.
- Creates a `purchaseDiscountedProduct` and `purchaseDiscountedPackage`. Since there is no overloads in JS, and upgrade info and type parameters in the regular purchase calls are only relevant to Android I figured it was cleaner to create new functions.
- Added a WIP `DEVELOPMENT.md` to keep track of how to use this repo during development.
- Moved `introPrice` into its own `PurchasesIntroPrice` object. The object is the same as the discounts used for the subscription offer but without `identifier`. I thought it was cleaner to just put it in its own object.

### CODE REVIEW

File: example/app/screens/UpsellScreen.js [hunk 1]
```diff
@@ -105,9 +105,10 @@ export default class UpsellScreen extends React.Component {
               <Button
                 color="#f2545b"
                 onPress={async () => {
-                  const aPackage = this.state.offerings.current.lifetime;
+                  const aPackage = this.state.offerings.current.monthly;
                   try {
-                    const purchaseMade = await Purchases.purchasePackage(aPackage, {oldSKU: "old", prorationMode: Purchases.PRORATION_MODE.DEFERRED}, Purchases.PURCHASE_TYPE.SUBS);
+                    const paymentDiscount = await Purchases.getPaymentDiscount(aPackage)
+                    const purchaseMade = await Purchases.purchasePackage(aPackage, {oldSKU: "old", prorationMode: Purchases.PRORATION_MODE.DEFERRED}, paymentDiscount);
```
Comment (vegaro) [inline example/app/screens/UpsellScreen.js]: remove this example

File: __tests__/index.test.js [hunk 2]
```diff
@@ -526,6 +526,125 @@ describe("Purchases", () => {
...
+      discounts: [discountStub]
+    }
+
+    Purchases.getPaymentDiscount(aProduct, discountStub)
+
+    expect(NativeModules.RNPurchases.getPaymentDiscount).toBeCalledWith(aProduct.identifier, discountStub.identifier);
+    expect(NativeModules.RNPurchases.getPaymentDiscount).toBeCalledTimes(1);
+  });
+
+  it("getPaymentDiscount returns undefined for Android", async () => {
```
Comment (aboedo) [inline __tests__/index.test.js]: minor, but the standard syntax for these is supposed to read sort of like a phrase, like when they fail:
"when getting payment discount it returns undefined for android failed"
would be expressed as 

```javascript
context("getting payment discount", () => { 
    it ("returns undefined for Android", () => {
```

see https://github.com/rubocop-hq/rspec-style-guide

File: ios/RNPurchases.m [hunk 3]
```diff
@@ -43,6 +43,7 @@ - (dispatch_queue_t)methodQueue
 {
     [RCPurchases configureWithAPIKey:apiKey appUserID:appUserID observerMode:observerMode];
     RCPurchases.sharedPurchases.delegate = self;
+    [RCCommonFunctionality initialize];
```
Comment (aboedo) [inline ios/RNPurchases.m]: `configure`?

File: ios/RNPurchases.m [hunk 4]
```diff
@@ -56,8 +57,8 @@ - (dispatch_queue_t)methodQueue
     [RCCommonFunctionality setFinishTransactions:finishTransactions];
 }
 
-RCT_REMAP_METHOD(addAttributionData, 
-                 addAttributionData:(NSDictionary *)data 
+RCT_REMAP_METHOD(addAttributionData,
```
Comment (aboedo) [inline ios/RNPurchases.m]: is remap needed here? you're keeping the name the same, right?

File: ios/RNPurchases.m [hunk 5]
```diff
@@ -85,21 +86,23 @@ - (dispatch_queue_t)methodQueue
                  purchaseProduct:(NSString *)productIdentifier
                  upgradeInfo:(NSDictionary *)upgradeInfo
                  type:(NSString *)type
+                 signedDiscountTimestamp:(NSString *)signedDiscountTimestamp
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)
 {
-    [RCCommonFunctionality purchaseProduct:productIdentifier completionBlock:[self getResponseCompletionBlockWithResolve:resolve reject:reject]];
+    [RCCommonFunctionality purchaseProduct:productIdentifier signedDiscountTimestamp:signedDiscountTimestamp completionBlock:[self getResponseCompletionBlockWithResolve:resolve reject:reject]];
```
Comment (aboedo) [inline ios/RNPurchases.m]: nitpick: breaking line between params here makes it easier to read without horizontal scrolling

Comment (aboedo) [reply ios/RNPurchases.m]: same for other similar calls

File: __tests__/index.test.js (hunk 2 above)
Comment (vegaro) [reply __tests__/index.test.js]: oh cool

File: ios/RNPurchases.m (hunk 3 above)
Comment (vegaro) [reply ios/RNPurchases.m]: yeah forgot to push

**Comment by vegaro** [comment]
@aboedo ready 🙏

**Pull request merged**